IdentityCore/src/throttling/thumbprint_calculation/MSIDThumbprintCalculator.m

```diff
@@ -0,0 +1,104 @@
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.  
+
+
+#import <Foundation/Foundation.h>
+#import "MSIDThumbprintCalculator.h"
+
+
+//Exclude List:
+//1) Client ID - same across all requests
+//2) Grant type - fixed as @"refresh_token"
+@implementation MSIDThumbprintCalculator
+
++ (NSString *)calculateThumbprint:(NSDictionary *)requestParameters
+                     filteringSet:(NSSet *)filteringSet
+                shouldIncludeKeys:(BOOL)shouldIncludeKeys
+{
+    NSArray *sortedThumbprintRequestList = [self sortRequestParametersUsingFilteredSet:requestParameters
+                                                                          filteringSet:filteringSet
+                                                                     shouldIncludeKeys:shouldIncludeKeys];
+    if (sortedThumbprintRequestList)
+    {
+        NSUInteger thumbprintKey = [self hash:sortedThumbprintRequestList];
+        if (thumbprintKey == 0)
+        {
+            //Log Warning
+            return nil;
+        }
+        
+        else
+        {
+            return [NSString stringWithFormat:@"%lu", thumbprintKey];
+        }
+    }
+    return nil;
+}
+
++ (NSArray *)sortRequestParametersUsingFilteredSet:(NSDictionary *)requestParameters
+                                      filteringSet:(NSSet *)filteringSet
+                                 shouldIncludeKeys:(BOOL)shouldIncludeKeys
+{
+    NSMutableArray *arrayList = [NSMutableArray new];
+    [requestParameters enumerateKeysAndObjectsUsingBlock:^(id  _Nonnull key, id  _Nonnull obj, __unused BOOL * _Nonnull stop) {
```

Comment:
I think NSPredicate works better if the input container is of NSArray type. You would first need to convert input dictionary into nsarray object and then pass in predicatewithformat or predicatewithblock, which could be more complex, especially if dictionary key values are all different. 
For NSDictionary, the only way to use NSPredicate directly on it is by first turning it into array object either by keys or by values as below:
NSArray *resultArray = [[mainDict allValues] filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"(key1 == %@) AND (key2==%@)", @"tt",@"vv"]];

In most cases, predicatewithFormat tries to compare by expected key values, while for our case, we have variable set of key values. (client_id, enrollment_id, etc.) 

Lastly, predicate statement becomes very complex when checking for compound conditions, as we are doing here. 

**So I think it would be easier to stay with enumerateKeysandObjectsUsingBlock.** 

Resources I used to reach above conclusion: 
https://stackoverflow.com/questions/16556905/filtering-nsdictionary-with-predicate
https://medium.com/@mykelp/pretty-powerful-predicates-cfb29c3493fb
https://nshipster.com/nspredicate/

Reply:
Sounds good!